### ISSUE TITLE
fixtures - 

### ISSUE THREAD

**Issue by ltalirz** [aiidateam/aiida-core#3473]
aiida-core 1.0 now ships with a number of fixtures for running tests, but there is clearly room for improvement.

We should decide which of the following features should also be included:

fixtures by @sphuber in aiida-qe:

* There is a fixture to test [`CalcJob`](https://github.com/aiidateam/aiida-quantumespresso/blob/89172686a824d0838e0da2c017bf793184aa92ce/tests/conftest.py#L59) classes. Example usage can be found here for [`PwCalculation`](https://github.com/aiidateam/aiida-quantumespresso/blob/develop/tests/calculations/test_pw.py).
* Similar fixture exists to test [`Parser`](https://github.com/aiidateam/aiida-quantumespresso/blob/89172686a824d0838e0da2c017bf793184aa92ce/tests/conftest.py#L86) implementations. It essentially mocks a completed `CalcJobNode` with the `retrieved` folder data node that contains the retrieved files. Example usage for [`PwParser`](https://github.com/aiidateam/aiida-quantumespresso/blob/89172686a824d0838e0da2c017bf793184aa92ce/tests/parsers/test_pw.py#L77) shows.

The automatic comparison files generation and checking is done through the [`pytest-regressions`](https://github.com/aiidateam/aiida-quantumespresso/blob/89172686a824d0838e0da2c017bf793184aa92ce/setup.json#L65) plugin which is added as a dependency. With this pytest plugin and the two fixtures, it is really easy to write and run tests for calculation jobs and parsers without actually having to run the calculations. Note of course, as mentioned before, that some tests that actually run the code will still be useful and desirable.

pointers from @greschd concerning aiida-pytest:

* example for the config file: https://github.com/greschd/aiida-tbextraction/blob/master/tests/config_example.yml
  On the actual machine I would rename this to ‘config.yml’ (which is in .gitignore), and make the necessary changes.
 * Getting the config: https://github.com/greschd/aiida_pytest/blob/migrate_v1/aiida_pytest/_configure.py
 * Setup of computers: https://github.com/greschd/aiida_pytest/blob/migrate_v1/aiida_pytest/_computer.py
 * Setup of codes: https://github.com/greschd/aiida_pytest/blob/migrate_v1/aiida_pytest/_code.py

One thing that I’ve found to be helpful is having an option to wait for user input before wiping the test DB. That allows for inspecting the state better before it’s gone. See here: https://github.com/greschd/aiida_pytest/blob/migrate_v1/aiida_pytest/_configure.py#L152

**Comment by greschd** [comment]
Another thing I had forgot to mention: It would be great to print the ``AIIDA_PATH`` during setup -- again to allow inspecting the test DB.

I think ``pytest`` has some way to allow for printing that _before_ the tests start, but I haven't looked into that.

**Comment by chrisjsewell** [comment]
> I think pytest has some way to allow for printing before the tests start

FYI the `aiida-crystal17` [conftest](https://github.com/chrisjsewell/aiida-crystal17/blob/develop/conftest.py) has some of the more advanced features of `pytest`, like printing a header (`pytest_report_header`), setting up some command-line and configuration file options, and running selective test sets.

**Comment by greschd** [comment]
This might require that we define the ``root_dir`` already at an earlier time (before the fixtures are being set up), and then pass it on to the ``TestManager``. Shouldn't be a problem though - or is the ``root_dir`` supposed to be changeable between different tests?

**Comment by ltalirz** [comment]
>  or is the root_dir supposed to be changeable between different tests?

Haven't yet seen a use case for this, so not at the moment.

**Comment by espenfl** [comment]
Yes, so we should probably make something like a `aiida_profile.root_dir` or similar, so that we do not have to use the protected `aiida_profile._manager.root_dir`. Second, we introduced the posibility to pass arguments to `pgtest` (needed for instance to set the `pg_ctl` path if you have multiple Postgresql installations) so we need to open for that again. Easiest I guess is just to make the fixture parameterized.

**Comment by espenfl** [comment]
Also, if there is a feature to test `CalcJob` I would think the extension to `WorkChain` should be rather straightforward? We have been running mock executables here: https://github.com/aiida-vasp/aiida-vasp/blob/develop/aiida_vasp/commands/mock_vasp.py which is triggered from a `WorkChain` test as: https://github.com/aiida-vasp/aiida-vasp/blob/develop/aiida_vasp/workchains/tests/test_vasp_wc.py

This is rather clean and easy. However, gives some problems during development when tracking down where things fail (the combination of commands in shell/script and quite a lot of files makes this a bit hard to track). In particular it gets a bit nasty when we for instance want to test something like a k-point convergence `WorkChain`, like here: https://github.com/aiida-vasp/aiida-vasp/blob/develop/aiida_vasp/workchains/tests/test_converge_wc.py where we need to pass an argument that gives an indication where to pull the files. I think there is no way around this as we need the files to perform realistic tests, but it would be nice to have a more formal context to do this. Also, in particular a context that makes it easier to eject sensible error messages when the test fails etc.

**Comment by espenfl** [comment]
> Yes, so we should probably make something like a `aiida_profile.root_dir` or similar, so that we do not have to use the protected `aiida_profile._manager.root_dir`. Second, we introduced the posibility to pass arguments to `pgtest` (needed for instance to set the `pg_ctl` path if you have multiple Postgresql installations) in #2941 so we need to open for that again. Easiest I guess is just to make the fixture parameterized.

In order to introduce the `pgtest` passing, I have opened a WIP PR #3481 on this where we can discuss this issue in detail.

**Comment by danielhollas** [comment]
> Second, we introduced the posibility to pass arguments to pgtest (needed for instance to set the pg_ctl path if you have multiple Postgresql installations) in https://github.com/aiidateam/aiida-core/pull/2941 so we need to open for that again. Easiest I guess is just to make the fixture parameterized.

> In order to reintroduce the pgtest argument passing, I have opened a WIP PR https://github.com/aiidateam/aiida-core/pull/3481 on this where we can discuss this issue in detail.
>My vote going forward is to open for passing options to pgtest in a config file.

This is now hopefully less pressing now that the new fixtures in `aiida.tools.pytest_fixtures` use the `sqlite_dos` backend by default.

**Comment by mbercx** [comment]
I'm currently working on moving some plugins out of `aiida-quantumespresso`, and indeed realised (again) that several test fixtures should (and already do) have a home in `aiida-core`, but aren't available in `aiida.tools.pytest_fixtures`. I'll get the ball rolling here by making e.g. `generate_calc_job` importable and using it in `aiida_quantumespresso`.
